pkg/controller/controller_instance.go

```diff
@@ -212,1168 +211,520 @@ func (c *controller) reconcileServiceInstanceKey(key string) error {
 	return c.reconcileServiceInstance(instance)
 }
 
-// reconcileServiceInstanceDelete is responsible for handling any instance whose
-// deletion timestamp is set.
-func (c *controller) reconcileServiceInstanceDelete(instance *v1beta1.ServiceInstance) error {
-	// nothing to do...
-	if instance.DeletionTimestamp == nil && !instance.Status.OrphanMitigationInProgress {
-		return nil
+// reconcileServiceInstance is the control-loop for reconciling Instances. An
+// error is returned to indicate that the instance has not been fully
+// processed and should be resubmitted at a later time.
+func (c *controller) reconcileServiceInstance(instance *v1beta1.ServiceInstance) error {
+	switch {
+	case instance.Status.AsyncOpInProgress:
+		return c.pollServiceInstance(instance)
+	case instance.ObjectMeta.DeletionTimestamp != nil || instance.Status.OrphanMitigationInProgress:
+		return c.reconcileServiceInstanceDelete(instance)
+	case instance.Status.ReconciledGeneration != 0:
+		return c.reconcileServiceInstanceUpdate(instance)
+	default:
```

Comment:
I think it reads better if the default is do nothing, can there be a case to know if it is an add operation?

Reply:
Hmm... I don't think there'd be a real win by separating it out, because the case would be `instance.Status.ReconciledGeneration == 0`, which is just an else-dup of the Update one.

Reply:
how about that in a comment?